pkg/urbit/vere/io/ames.c

```diff
@@ -435,6 +550,112 @@ _ames_hear_bail(u3_ovum* egg_u, u3_noun lud)
   u3_ovum_free(egg_u);
 }
 
+/* _ames_put_packet(): add packet to queue, drop old packets on pressure
+*/
+static void
+_ames_put_packet(u3_ames* sam_u,
+                 u3_noun  msg,
+                 u3_noun  lan)
+{
+  u3_noun wir = u3nc(c3__ames, u3_nul);
+  u3_noun cad = u3nt(c3__hear, u3nc(c3n, lan), msg);  // trnasfer, unchanged
+
+  u3_auto_peer(
+    u3_auto_plan(&sam_u->car_u,
+                 u3_ovum_init(0, c3__a, wir, cad)),
+    0, 0, _ames_hear_bail);
+
+  _ames_cap_queue(sam_u);
+}
+
+/*  _ames_forward(): forward pac_u onto lan, then free pac_u
+*/
+static void
+_ames_forward(u3_panc* pac_u, u3_noun lan)
+{
+  c3_o nal = c3n;
+
+  //  repack the maybe-updated packet
+  //
+  {
+    //  unpack (jam [(unit lane) body])
+    //
+    u3_noun lon, bod;
+    {
+      u3_noun old = u3ke_cue(pac_u->bod_u.con);
+      u3x_cell(old, &lon, &bod);
+      u3k(lon); u3k(bod);
+      u3z(old);
+    }
+
+    //  only replace the lane if it was ~
+    //
+    if (u3_nul == lon) {
+      u3z(lon);
+      lon = u3nt(u3_nul, c3n, u3k(pac_u->ore));
+      //TODO  ore is always opaque lane. what about the sender=galaxy case?
+      nal = c3y;
+    }
+
+    pac_u->bod_u.con = u3ke_jam(u3nc(lon, bod));
+  }
+
+  //  if we updated the origin lane, serialize the body to update the mug hash
+  //
+  //TODO  refactor into serialize_packet?
+  if (c3y == nal) {
+    c3_w  bod_w = u3r_met(3, pac_u->bod_u.con);
+    c3_y  sen_y = 2 << pac_u->hed_u.sac_y;
+    c3_y  rec_y = 2 << pac_u->hed_u.rac_y;
+    c3_y* bod_y = c3_malloc(sen_y + rec_y + bod_w);
+    u3r_bytes(0, sen_y, bod_y,                 pac_u->bod_u.sen);
+    u3r_bytes(0, rec_y, bod_y + sen_y,         pac_u->bod_u.rec);
+    u3r_bytes(0, bod_w, bod_y + sen_y + rec_y, pac_u->bod_u.con);
+    u3_noun bod = u3i_bytes(sen_y + rec_y + bod_w, bod_y);
+
+    pac_u->hed_u.mug_l = u3r_mug(bod) & ((1 << 20) - 1);
+
+    c3_free(bod_y);
+    u3z(bod);
+  }
+
+  pac_u->sam_u->fow_d++;
+  if ( 0 == (pac_u->sam_u->fow_d % 1000) ) {
+    u3l_log("ames: forwarded %" PRIu64 " total\n", pac_u->sam_u->fow_d);
```

Comment:
This seems like it would print a lot on a live infrastructure node.  Maybe decrease the frequency?